File: frontend/src/utils/pagination.ts
import type { PageResult } from '@/types/domain'

export const OPTION_PAGE_SIZE = 100

type PageLoader<T> = (params: { page: number; page_size: number }) => Promise<PageResult<T>>

export async function fetchAllPages<T>(loader: PageLoader<T>, pageSize = OPTION_PAGE_SIZE) {
  const items: T[] = []

  for (let page = 1; page <= 100; page += 1) {
    const result = await loader({ page, page_size: pageSize })
    const pageItems = Array.isArray(result.items) ? result.items : []
    items.push(...pageItems)

    const total = result.total || items.length
    const effectivePage = result.page || page
    const effectivePageSize = result.page_size || pageSize
    if (!pageItems.length || items.length >= total || effectivePage * effectivePageSize >= total) {
      break
    }
  }

  return items
}

File: internal/repository/pagination.go
package repository

type Page struct {
	Page     int
	PageSize int
}

func (p Page) Offset() int {
	if p.Page <= 1 {
		return 0
	}
	return (p.Page - 1) * p.Limit()
}

func (p Page) Limit() int {
	if p.PageSize <= 0 {
		return 20
	}
	if p.PageSize > 100 {
		return 100
	}
	return p.PageSize
}
